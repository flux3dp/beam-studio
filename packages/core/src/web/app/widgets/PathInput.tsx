import React, { useEffect, useRef, useState } from 'react';
import classNames from 'classnames';

import dialog from 'implementations/dialog';
import fs from 'implementations/fileSystem';
import { OpenDialogProperties } from 'interfaces/IDialog';

import styles from './PathInput.module.scss';

export enum InputType {
  FILE = 0,
  FOLDER = 1,
  BOTH = 2,
}

const propertiesMap = {
  0: ['openFile'],
  1: ['openDirectory', 'createDirectory', 'promptToCreate'],
  2: ['openFile', 'openDirectory', 'createDirectory', 'promptToCreate'],
};

interface Props {
  buttonTitle: string;
  className?: string;
  defaultValue: string;
  forceValidValue?: boolean;
  type: InputType;
  getValue: (val: string, isValid: boolean) => void;
}

const PathInput = ({
  buttonTitle,
  className,
  defaultValue = '',
  getValue,
  forceValidValue = true,
  type,
}: Props): JSX.Element => {
  const [displayValue, setDisplayValue] = useState(defaultValue);
  const [savedValue, setSavedValue] = useState(defaultValue);
  const inputEl = useRef(null);

  useEffect(() => {
    setDisplayValue(defaultValue);
    setSavedValue(defaultValue);
  }, [defaultValue]);

  const validateValue = (val: string) => {
    if (fs.exists(val)) {
      if (type === InputType.BOTH) return true;
      return (
        (type === InputType.FILE && fs.isFile(val)) ||
        (type === InputType.FOLDER && fs.isDirectory(val))
      );
    }
    return false;
  };

  const updateValue = () => {
    const isValid = validateValue(displayValue);
    if (!forceValidValue || isValid) {
      if (displayValue !== savedValue) {
        setSavedValue(displayValue);
        getValue(displayValue, isValid);
      }
    } else {
      setDisplayValue(savedValue);
    }
  };

  const handleChange = (e: React.ChangeEvent<HTMLInputElement>) => {
    const target = e.target as HTMLInputElement;
    setDisplayValue(target.value);
  };

  const handleKeyUp = (e: React.KeyboardEvent) => {
    // backspace somehow does not trigger onchange event
    const target = e.target as HTMLInputElement;
    if (target.value !== displayValue) {
      setDisplayValue(target.value);
    }
  };

  const setValueFromDialog = async () => {
    const properties = propertiesMap[type] as OpenDialogProperties[];
    const option = {
      properties,
      defaultPath: savedValue,
    };
    const { filePaths, canceled } = await dialog.showOpenDialog(option);
    if (!canceled) {
      const isValid = validateValue(filePaths[0]);
      if (!forceValidValue || isValid) {
        setSavedValue(filePaths[0]);
        setDisplayValue(filePaths[0]);
        getValue(filePaths[0], isValid);
      }
    }
  };

  return (
    <div className={classNames(styles.container, className)}>
      <div className={styles.btn} title={buttonTitle} onClick={setValueFromDialog}>
        <img src="img/right-panel/icon-import.svg" />
      </div>
      <input
        id="location-input"
        type="text"
        value={displayValue}
        onBlur={updateValue}
        onChange={handleChange}
        onKeyUp={handleKeyUp}
        ref={inputEl}
      />
    </div>
  );
};

export default PathInput;
