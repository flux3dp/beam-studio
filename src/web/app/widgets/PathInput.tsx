import electronDialogs from 'app/actions/electron-dialogs';
import { IFileFilter } from 'interfaces/IElectron';

const classNames = requireNode('classnames');
const React = requireNode('react');
const fs = requireNode('fs');
const fsPromises = fs.promises;

const { useEffect, useRef, useState } = React;

export enum InputType {
  FILE = 0,
  FOLDER = 1,
  BOTH = 2,
};

const propertiesMap = {
  0: ['openFile',],
  1: ['openDirectory', 'createDirectory', 'promptToCreate'],
  2: ['openFile', 'openDirectory', 'createDirectory', 'promptToCreate'],
}

interface IProps {
  buttonTitle?: string,
  className?: string
  defaultValue: string,
  getValue: Function,
  forceValidValue?: boolean,
  onBlur?: Function,
  type: InputType,
  filters?: IFileFilter[],
};

const PathInput = ({ buttonTitle, className, defaultValue, getValue, forceValidValue = true, onBlur, type = InputType.FILE, filters, }: IProps) => {
  const [displayValue, setDisplayValue] = useState(defaultValue);
  const [savedValue, setSavedValue] = useState(defaultValue);
  const inputEl = useRef(null);
  useEffect(() => {
    setDisplayValue(defaultValue);
    setSavedValue(defaultValue);
  }, [defaultValue]);

  const handleBlur = (e) => {
    updateValue();
    if (onBlur) {
      onBlur();
    }
  };
  const handleChange = (e: InputEvent) => {
    const target = e.target as HTMLInputElement;
    setDisplayValue(target.value);
  };
  const handleKeyUp = (e: KeyboardEvent) => {
    // backspace somehow does not trigger onchange event
    const target = e.target as HTMLInputElement;
    if (target.value !== displayValue) {
      setDisplayValue(target.value);
    }
  }

  const validateValue = (val: string) => {
    if (fs.existsSync(val)) {
      if (type === InputType.BOTH) return true;
      const stat = fs.lstatSync(val);
      return (type === InputType.FILE && stat.isFile()) || (type === InputType.FOLDER && stat.isDirectory());
    }
    return false;
  }

  const updateValue = () => {
    const isValid = validateValue(displayValue);
    if (!forceValidValue || isValid) {
      if (displayValue !== savedValue) {
        setSavedValue(displayValue);
        if (getValue) {
          getValue(displayValue, isValid);
        }
      }
    } else {
      setDisplayValue(savedValue);
    }
  }

  const setValueFromDialog = async () => {
    const properties = propertiesMap[type];
    const option = {
      properties,
      filters,
      defaultPath: savedValue,
    };
    const { filePaths, canceled } = await electronDialogs.showOpenDialog(option);
    if (!canceled) {
      const isValid = validateValue(filePaths[0]);
      if (!forceValidValue || isValid) {
        setSavedValue(filePaths[0]);
        setDisplayValue(filePaths[0]);
        if (getValue) {
          getValue(filePaths[0], isValid);
        }
      }
    }
  }

  const openDialogButton = (
    <div className='dialog-btn' title={buttonTitle} onClick={setValueFromDialog}>
      <img src={'img/right-panel/icon-import.svg'} />
    </div>
  );

  return (
    <div className={classNames('path-input', className)}>
      <input
        type='text'
        value={displayValue}
        onBlur={handleBlur}
        onChange={handleChange}
        onKeyUp={handleKeyUp}
        ref={inputEl}
      />
      {openDialogButton}
    </div>
  );
}

export default PathInput;