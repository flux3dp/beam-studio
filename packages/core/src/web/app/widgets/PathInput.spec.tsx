/* eslint-disable import/first */
import React from 'react';
import { act, fireEvent, render } from '@testing-library/react';

const mockShowOpenDialog = jest.fn();
jest.mock('implementations/dialog', () => ({
  showOpenDialog: mockShowOpenDialog,
}));

const mockExists = jest.fn();
const mockIsFile = jest.fn();
const mockIsDirectory = jest.fn();
jest.mock('implementations/fileSystem', () => ({
  exists: mockExists,
  isFile: mockIsFile,
  isDirectory: mockIsDirectory,
}));

import PathInput from './PathInput';

describe('test PathInput', () => {
  test('should render correctly', async () => {
    const mockGetValue = jest.fn();
    const { container } = render(
      <PathInput
        buttonTitle="Choose Folder"
        className="with-error"
        defaultValue="defaultFolder"
        forceValidValue={false}
        getValue={mockGetValue}
        type={1}
      />
    );
    expect(container).toMatchSnapshot();
    mockShowOpenDialog.mockResolvedValue({
      filePaths: [],
      canceled: true,
    });
    await act(async () => {
      fireEvent.click(container.querySelector('.btn'));
    });
    expect(mockShowOpenDialog).toHaveBeenCalledTimes(1);
    expect(mockShowOpenDialog).toHaveBeenNthCalledWith(1, {
      properties: ['openDirectory', 'createDirectory', 'promptToCreate'],
      defaultPath: 'defaultFolder',
    });
    expect(mockGetValue).not.toHaveBeenCalled();
    expect(container).toMatchSnapshot();

    mockShowOpenDialog.mockResolvedValue({
      filePaths: ['myDocuments'],
      canceled: false,
    });
    mockExists.mockReturnValue(false);
    await act(async () => {
      fireEvent.click(container.querySelector('.btn'));
    });
    expect(mockShowOpenDialog).toHaveBeenCalledTimes(2);
    expect(mockShowOpenDialog).toHaveBeenNthCalledWith(2, {
      properties: ['openDirectory', 'createDirectory', 'promptToCreate'],
      defaultPath: 'defaultFolder',
    });
    expect(mockExists).toHaveBeenCalledTimes(1);
    expect(mockExists).toHaveBeenNthCalledWith(1, 'myDocuments');
    expect(mockGetValue).toHaveBeenCalledTimes(1);
    expect(mockGetValue).toHaveBeenNthCalledWith(1, 'myDocuments', false);
    expect(container).toMatchSnapshot();
  });
});
