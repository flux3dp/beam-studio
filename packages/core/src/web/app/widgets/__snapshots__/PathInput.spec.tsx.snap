// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test PathInput should render correctly 1`] = `
<div>
  <div
    class="container with-error"
  >
    <div
      class="btn"
      title="Choose Folder"
    >
      <img
        src="img/right-panel/icon-import.svg"
      />
    </div>
    <input
      id="location-input"
      type="text"
      value="defaultFolder"
    />
  </div>
</div>
`;

exports[`test PathInput should render correctly 2`] = `
<div>
  <div
    class="container with-error"
  >
    <div
      class="btn"
      title="Choose Folder"
    >
      <img
        src="img/right-panel/icon-import.svg"
      />
    </div>
    <input
      id="location-input"
      type="text"
      value="defaultFolder"
    />
  </div>
</div>
`;

exports[`test PathInput should render correctly 3`] = `
<div>
  <div
    class="container with-error"
  >
    <div
      class="btn"
      title="Choose Folder"
    >
      <img
        src="img/right-panel/icon-import.svg"
      />
    </div>
    <input
      id="location-input"
      type="text"
      value="myDocuments"
    />
  </div>
</div>
`;
